### PR TITLE
Adding subject with tasks

### PR DESCRIPTION
**Pull request by Rownaknira**

### CODE REVIEW

File: app/controllers/subjects_controller.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
+class SubjectsController < ApplicationController
+  before_action :load_subject, only: [:edit, :update, :destroy]
+
+  def index
+    @subjects = Subject.all
+  end
+
+  def new
+    @subject = Subject.new
+    @task = @subject.tasks.build
```
Comment (abdullahallmehedi) [inline app/controllers/subjects_controller.rb]: why we need @task

File: app/controllers/subjects_controller.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  end
+
+  def destroy
+    @subject.destroy
+    redirect_to subjects_url
+  end
+
+  private
+  def subject_params
+    params.require(:subject).permit :id, :name, :description, tasks_attributes: [ :id, :name, :subject_id]
```
Comment (abdullahallmehedi) [inline app/controllers/subjects_controller.rb]: more than 80 characters

File: app/helpers/subjects_helper.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
+module SubjectsHelper
+  def link_to_add_fields(name, f, association, cssClass, title)  
+    new_object = f.object.class.reflect_on_association(association).klass.new
+    fields = f.fields_for(association, new_object, :child_index => "new_#{association}") do |builder|
```
Comment (abdullahallmehedi) [inline app/helpers/subjects_helper.rb]: 80 characters + trailing space

File: app/helpers/subjects_helper.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
+module SubjectsHelper
+  def link_to_add_fields(name, f, association, cssClass, title)  
+    new_object = f.object.class.reflect_on_association(association).klass.new
+    fields = f.fields_for(association, new_object, :child_index => "new_#{association}") do |builder|  
+      render(association.to_s.singularize + "_fields", :f => builder)
+    end  
+    link_to name, "#", :onclick => h("add_fields(this, \"#{association}\", \"#{escape_javascript(fields)}\")"), class: cssClass, title: title, remote: true
```
Comment (abdullahallmehedi) [inline app/helpers/subjects_helper.rb]: 80 characters + onclick:

File: app/views/subjects/_task_fields.erb [hunk 5]
```diff
@@ -0,0 +1,12 @@
+<tr>
+  <td>
+    <div class="form-group">
+      <%= f.text_field :name, class: "form-control",placeholder: t('tasks.name_placeholder') %>
```
Comment (abdullahallmehedi) [inline app/views/subjects/_task_fields.erb]: 80 characters

File: app/views/subjects/edit.html.erb [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+        <%= f.label :tasks %>
+        <table>
+          <tbody>
+            <%= f.fields_for :tasks do |builder| %>
+              <%= render "task_fields", :f => builder %>
+            <% end %>
+          </tbody>
+        </table>
+        <div>
+          <%= link_to_add_fields t('views.add_more'), f, :tasks, "btn btn-primary", t('views.add_more') %>
```
Comment (abdullahallmehedi) [inline app/views/subjects/edit.html.erb]: 80 characters

File: app/views/subjects/_task_fields.erb [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+    </div>
+  </td>
+  <td>
+    <div class="form-group">
+      <%= link_to_remove_fields t('views.remove'), f %>
```
Comment (abdullahallmehedi) [inline app/views/subjects/_task_fields.erb]: All i18n need to be re-factored.

File: app/views/subjects/edit.html.erb [hunk 8]
```diff
@@ -0,0 +1,30 @@
+<h1><%=t('subject.edit_subj') %></h1>
```
Comment (abdullahallmehedi) [inline app/views/subjects/edit.html.erb]: space + i18n

File: app/views/subjects/edit.html.erb [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+      <div class="form-group">
+        <%= f.label :description %>
+        <%= f.text_area :description ,class: 'form-control'%>
+      </div>
+      <div class="field">
+        <%= f.label :tasks %>
+        <table>
+          <tbody>
+            <%= f.fields_for :tasks do |builder| %>
+              <%= render "task_fields", :f => builder %>
```
Comment (abdullahallmehedi) [inline app/views/subjects/edit.html.erb]: f: builder

File: app/views/subjects/index.html.erb [hunk 10]
```diff
@@ -0,0 +1,23 @@
+<div class="row">
+  <h1 class="text-capitalize"><%= t('subject.list')%></h1>
```
Comment (abdullahallmehedi) [inline app/views/subjects/index.html.erb]: i18n + spacing

File: app/views/subjects/index.html.erb [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+        <th>Description</th>
+        <th colspan="2"></th>
+      </tr>
+    </thead>
+    <% @subjects.each do |subject|%>
+      <tr>
+        <td><%= subject.name%></td>
+        <td><%= subject.description%></td>
+        <td><%= link_to t('views.edit'),edit_subject_path(subject)%></td>
+        <td><%= link_to t('views.delete'),subject_path(subject),method: :delete,data: {confirm: t('views.delete_confirm')}%></td>
```
Comment (abdullahallmehedi) [inline app/views/subjects/index.html.erb]: 80 characters, spacing ..

File: app/views/subjects/new.html.erb [hunk 12]
```diff
@@ -0,0 +1,32 @@
+<h1><%=t('subject.add_subj') %></h1>
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%=form_for @subject  do |f| %>
```
Comment (abdullahallmehedi) [inline app/views/subjects/new.html.erb]: spacing issue

File: app/views/subjects/new.html.erb [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+    <div class="form-group">
+      <%= f.label :description %>
+      <%= f.text_area :description ,class: 'form-control'%>
+    </div>
+      <div class="field">
+        <%= f.label :tasks %>
+        <table>
+          <tbody>
+            <%= f.fields_for :tasks do |builder| %>
+              <%= render "task_fields", :f => builder %>
```
Comment (abdullahallmehedi) [inline app/views/subjects/new.html.erb]: f: builder

File: app/views/subjects/new.html.erb [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+            <% end %>
+          </tbody>
+        </table>
+        <div>
+          <%= link_to_add_fields t('views.add_more'), f, :tasks, "btn btn-primary", t('views.add_more') %>
```
Comment (abdullahallmehedi) [inline app/views/subjects/new.html.erb]: 80 characters

File: config/locales/en.yml [hunk 15]
```diff
@@ -8,3 +8,21 @@ en:
...
+    add_subj: "Add Subject"
+    edit_subj: "Edit Subject"
+  views:
+    add_more: "Add a new task"
+    remove: "Remove a task"
+    edit: "Edit"
+    delete: "Delete"
+    delete_confirm: "Are you sure you want to delete ?"
+  form:
+    create: "Create new"
```
Comment (abdullahallmehedi) [inline config/locales/en.yml]: Please re-factor whole file using i18n Rails plugin

**Comment by rofiuddin** [comment]
ok

File: app/helpers/subjects_helper.rb [hunk 16]
```diff
@@ -0,0 +1,23 @@
+module SubjectsHelper
+  def link_to_add_fields(name, f, association, cssClass, title)
+    new_object = f.object.class.reflect_on_association(association).klass.new
+    fields = f.fields_for(association, new_object, :child_index =>
+      "new_#{association}") do |builder|
+      render(association.to_s.singularize + "_fields", :f => builder)
```
Comment (abdullahallmehedi) [inline app/helpers/subjects_helper.rb]: f: builder

File: app/helpers/subjects_helper.rb [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+    end
+    link_to name, "#", :onclick => h("add_fields(this, \"#{association}\",
```
Comment (abdullahallmehedi) [inline app/helpers/subjects_helper.rb]: onclick: h( ... ..... .. .

File: app/views/subjects/index.html.erb [hunk 18]
```diff
@@ -0,0 +1,24 @@
+<div class="row">
+  <h1 class="text-capitalize"><%= t('.list') %></h1>
+  <br>
+  <table class="table table-striped">
+    <thead>
+      <tr>
+        <th><%= t('.name') %></th>
+        <th><%= t('.desc') %></th>
```
Comment (zenith19) [inline app/views/subjects/index.html.erb]: t '.name'

File: app/helpers/subjects_helper.rb [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+      \"#{escape_javascript(fields)}\")"), class: cssClass, title: title,
+      remote: true
+  end
+
+  def link_to_remove_fields name, f
+    f.hidden_field(:_destroy) + link_to_function(name, "remove_fields(this)")
+  end
+
+  def link_to_function name, function, html_options={}
+    onclick = "#{"#{html_options[:onclick]}; " if
```
Comment (abdullahallmehedi) [inline app/helpers/subjects_helper.rb]: please use if block here .. not inline statement

File: app/views/subjects/index.html.erb [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+        <td><%= subject.name%></td>
+        <td><%= subject.description%></td>
+        <td><%= link_to t('.edit_link'), edit_subject_path(subject) %></td>
+        <td><%= link_to t('.delete_link'), subject_path(subject),
+          method: :delete, data: {confirm: t('.delete_confirm')} %></td>
+      </tr>
+    <% end %>
+  </table>
+  <br>
+  <%= link_to t('.create'),new_subject_path,class: "btn btn-primary" %>
```
Comment (zenith19) [inline app/views/subjects/index.html.erb]: space
<%= link_to t('.create'), new_subject_path, class: "btn btn-primary" %>

File: app/views/subjects/edit.html.erb [hunk 21]
```diff
@@ -0,0 +1,31 @@
+<h1><%= t('.edit') %></h1>
```
Comment (abdullahallmehedi) [inline app/views/subjects/edit.html.erb]: remove brackets

File: app/views/subjects/index.html.erb [hunk 22]
```diff
@@ -0,0 +1,24 @@
+<div class="row">
+  <h1 class="text-capitalize"><%= t('.list') %></h1>
```
Comment (abdullahallmehedi) [inline app/views/subjects/index.html.erb]: remove bracket

File: app/views/subjects/index.html.erb [hunk 23]
```diff
@@ -0,0 +1,24 @@
...
+  <br>
+  <table class="table table-striped">
+    <thead>
+      <tr>
+        <th><%= t('.name') %></th>
```
Comment (abdullahallmehedi) [inline app/views/subjects/index.html.erb]: remove bracket

File: app/views/subjects/index.html.erb (hunk 18 above)
Comment (abdullahallmehedi) [inline app/views/subjects/index.html.erb]: remove bracket

File: app/views/subjects/index.html.erb (hunk 20 above)
Comment (abdullahallmehedi) [inline app/views/subjects/index.html.erb]: spacing issue

File: app/views/subjects/new.html.erb [hunk 24]
```diff
@@ -0,0 +1,31 @@
+<h1><%= t('.add') %></h1>
```
Comment (abdullahallmehedi) [inline app/views/subjects/new.html.erb]: remove bracket

File: app/views/subjects/new.html.erb [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for @subject  do |f| %>
+    <div class="form-group">
+      <%= f.label :name %>
+      <%= f.text_field :name ,class: 'form-control'%>
```
Comment (abdullahallmehedi) [inline app/views/subjects/new.html.erb]: spacing issue

File: app/views/subjects/new.html.erb [hunk 26]
```diff
@@ -0,0 +1,31 @@
...
+    </div>
+    <div class="form-group">
+      <%= f.label :description %>
+      <%= f.text_area :description ,class: 'form-control'%>
```
Comment (abdullahallmehedi) [inline app/views/subjects/new.html.erb]: spacing issue

File: app/controllers/subjects_controller.rb [hunk 27]
```diff
@@ -0,0 +1,47 @@
...
+
+  def destroy
+    @subject.destroy
+    redirect_to subjects_url
+  end
+
+  private
+  def subject_params
+    params.require(:subject).permit :id, :name, :description,
+      tasks_attributes: [ :id, :name, :subject_id]
```
Comment (chuanhtuan) [inline app/controllers/subjects_controller.rb]: _destroy

**Comment by rofiuddin** [comment]
ok

File: config/routes.rb [hunk 28]
```diff
@@ -4,4 +4,7 @@
   get "home"    => "static_pages#home"
   get "about"   => "static_pages#about"
   get "contact" => "static_pages#contact"
+
+  get 'subjects/new'
```
Comment (chuanhtuan) [inline config/routes.rb]: remove

**Pull request merged**